homeassistant/components/statistics/sensor.py

```diff
@@ -277,73 +342,91 @@ def _purge_old(self):
+            )
+            self.attr[STAT_VARIANCE] = round(
+                statistics.variance(self.states), self._precision
+            )
+        else:
+            self.attr[STAT_STANDARD_DEVIATION] = STATE_UNKNOWN
+            self.attr[STAT_VARIANCE] = STATE_UNKNOWN
+
+        if states_count > self._quantile_intervals:
+            self.attr[STAT_QUANTILES] = [
+                round(quantile, self._precision)
+                for quantile in statistics.quantiles(
+                    self.states,
+                    n=self._quantile_intervals,
+                    method=self._quantile_method,
+                )
+            ]
+        else:
+            self.attr[STAT_QUANTILES] = STATE_UNKNOWN
+
+        if states_count == 0:
+            self.attr[STAT_MEAN] = STATE_UNKNOWN
+            self.attr[STAT_MEDIAN] = STATE_UNKNOWN
+            self.attr[STAT_TOTAL] = STATE_UNKNOWN
+            self.attr[STAT_MIN_VALUE] = self.attr[STAT_MAX_VALUE] = STATE_UNKNOWN
+            self.attr[STAT_MIN_AGE] = self.attr[STAT_MAX_AGE] = STATE_UNKNOWN
+            self.attr[STAT_CHANGE] = self.attr[STAT_AVERAGE_CHANGE] = STATE_UNKNOWN
+            self.attr[STAT_CHANGE_RATE] = STATE_UNKNOWN
+            return
+
+        self.attr[STAT_MEAN] = round(statistics.mean(self.states), self._precision)
+        self.attr[STAT_MEDIAN] = round(statistics.median(self.states), self._precision)
+
+        self.attr[STAT_TOTAL] = round(sum(self.states), self._precision)
+        self.attr[STAT_MIN_VALUE] = round(min(self.states), self._precision)
+        self.attr[STAT_MAX_VALUE] = round(max(self.states), self._precision)
+
+        self.attr[STAT_MIN_AGE] = self.ages[0]
+        self.attr[STAT_MAX_AGE] = self.ages[-1]
+
+        self.attr[STAT_CHANGE] = self.states[-1] - self.states[0]
+
+        self.attr[STAT_AVERAGE_CHANGE] = self.attr[STAT_CHANGE]
+        self.attr[STAT_CHANGE_RATE] = 0
+        if states_count > 1:
+            self.attr[STAT_AVERAGE_CHANGE] /= len(self.states) - 1
+
+            time_diff = (
+                self.attr[STAT_MAX_AGE] - self.attr[STAT_MIN_AGE]
+            ).total_seconds()
+            if time_diff > 0:
+                self.attr[STAT_CHANGE_RATE] = self.attr[STAT_CHANGE] / time_diff
+        self.attr[STAT_CHANGE] = round(self.attr[STAT_CHANGE], self._precision)
+        self.attr[STAT_AVERAGE_CHANGE] = round(
+            self.attr[STAT_AVERAGE_CHANGE], self._precision
+        )
+        self.attr[STAT_CHANGE_RATE] = round(
+            self.attr[STAT_CHANGE_RATE], self._precision
+        )
+
     async def async_update(self):
         """Get the latest data and updates the states."""
         _LOGGER.debug("%s: updating statistics", self.entity_id)
-        if self._max_age is not None:
+        if self._samples_max_age is not None:
             self._purge_old()
 
-        self.count = len(self.states)
-
-        if not self.is_binary:
-            try:  # require only one data point
```

Comment:
How come we don't need the try/except anymore?

Reply:
Because the only reason these will throw `StatisticsError` (instead of ValueError or TypeError) is when there are not enough samples in the buffer. I've made this implicit condition explicit, which makes the whole function cleaner and less nested.